design/graph-manifest.md

```diff
@@ -0,0 +1,219 @@
+The reader should be able to tell by the title, and the opening paragraph, if this document is relevant to them.
+
+Currently, Velero does not have a complete manifest of everything in the backup, aside from the backup tarball itself.
+This change introduces a new data structure to be stored with a backup in object storage which will allow for more efficient operations in reporting of what a backup contains.
+Additionally, this manifest should enable advancements in Velero's features and architecture, enabling dry-run support, concurrent backup and restore operations, and reliable restoration of complex applications.
+
+## Background
+
+Right now, Velero backs up items one at a time, sorted by API Group and namespace.
+It also restores items one at a time, using the restoreResourcePriorities flag to indicate which order API Groups should have their objects restored first.
+While this does work currently, it presents challenges for more complex applications that have their dependencies in the form of a graph rather than strictly linear.
+
+For example, Cluster API clusters are a set of complex Kubernetes objects that require that the "root" objects are restored first, before their "leaf" objects.
+If a Cluster that a ClusterResourceSetBinding refers to does not exist, then a restore of the CAPI cluster will fail.
+
+Additionally, Velero does not have a reliable way to communicate what objects will be affected in a backup or restore operation without actually performing the operation.
+This complicates dry-run tasks, because a user must simply perform the action without knowing what will be touched.
+It also complicates allowing backups and restores to run in parallel, because there is currently no way to know if a single Kubernetes object is included in multiple backups or restores, which can lead to unreliability, deadlocking, and race conditions were Velero made to be more concurrent today.
+
+## Goals
+
+- Introduce a manifest data structure that defines the contents of a backup.
+- Store the manifest data into object storage alongside existing backup data.
+
+## Non Goals
+
+This proposal seeks to enable, but not define, the following.
+
+- Implementing concurrency beyond what already exists in Velero.
+- Implementing a dry-run feature.
+- Implementing a new restore ordering procedure.
+
+While the data structure should take these scenarios into account, they will not be implemented alongside it.
+
+## High-Level Design
+
+To uniquely identify a Kubernetes object within a cluster or backup, the following fields are sufficient:
+
+- API Group and Version (example: backup.velero.io/v1)
+- Namespace 
+- Name
+- Labels
+
+This criteria covers the majority of Velero's inclusion or exclusion logic.
+However, some additional fields enable further use cases.
+
+- Owners, which are other Kubernetes objects that have some relationship to this object. They may be strict or soft dependencies.
```

Comment:
There are objects that have relationships not called `owner`.

To properly handle these, I can think of 2 solutions

1) Use the definition (CR or Kubernetes in-tree) of the field to validate whether or not it's an object reference
2) Allow developers to supply an annotation on the CR describing the fields that are references.

I would prefer 1, but need to do a little research here.